cmd/runner/testcase.go

```diff
@@ -0,0 +1,225 @@
+}
+
+type testCase interface {
+	setup() error
+	run(client endpoint, server endpoint) error
+	verify() error
+}
+
+type testError struct {
+	err         string
+	funcName    string
+	unsupported bool
+}
+
+func (e *testError) Error() string {
+	if e.unsupported {
+		return fmt.Sprintf("Unsupported,%s(): %s", e.funcName, e.err)
+	}
+	return fmt.Sprintf("Failure,%s(): %s", e.funcName, e.err)
+}
+
+type testCaseDC struct {
+	name      string
+	timeout   time.Duration
+	outputDir string
+}
+
+func (t *testCaseDC) setup() error {
+	err := os.MkdirAll(testInputsDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	err = os.RemoveAll(testOutputsDir)
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(testOutputsDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	var inputParams strings.Builder
+
+	rootSignatureAlgorithm := utils.SignatureECDSAWithP521AndSHA512
+	err = utils.MakeRootCertificate(
+		&utils.Config{
+			Hostnames:          []string{"root.com"},
+			ValidFrom:          time.Now(),
+			ValidFor:           365 * 25 * time.Hour,
+			SignatureAlgorithm: rootSignatureAlgorithm,
+		},
+		filepath.Join(testInputsDir, "root.crt"),
+		filepath.Join(testInputsDir, "root.key"),
+	)
+	if err != nil {
+		return err
+	}
+	inputParams.WriteString(fmt.Sprintf("Root certificate algorithm: 0x%X\n", rootSignatureAlgorithm))
+
+	intermediateSignatureAlgorithm := utils.SignatureECDSAWithP256AndSHA256
+	err = utils.MakeIntermediateCertificate(
+		&utils.Config{
+			Hostnames:          []string{"example.com"},
+			ValidFrom:          time.Now(),
+			ValidFor:           365 * 25 * time.Hour,
+			SignatureAlgorithm: intermediateSignatureAlgorithm,
+			ForDC:              true,
+		},
+		filepath.Join(testInputsDir, "root.crt"),
+		filepath.Join(testInputsDir, "root.key"),
+		filepath.Join(testInputsDir, "example.crt"),
+		filepath.Join(testInputsDir, "example.key"),
+	)
+	if err != nil {
+		return err
+	}
+	inputParams.WriteString(fmt.Sprintf("Intermediate certificate algorithm: 0x%X\n", intermediateSignatureAlgorithm))
+
+	dcAlgorithm := utils.SignatureECDSAWithP256AndSHA256
+	dcValidFor := 24 * time.Hour
+	err = utils.MakeDelegatedCredential(
+		&utils.Config{
+			ValidFor:           dcValidFor,
+			SignatureAlgorithm: dcAlgorithm,
+		},
+		&utils.Config{},
+		filepath.Join(testInputsDir, "example.crt"),
+		filepath.Join(testInputsDir, "example.key"),
+		filepath.Join(testInputsDir, "dc.txt"),
+	)
+	if err != nil {
+		return err
+	}
+	inputParams.WriteString(fmt.Sprintf("Delegated credential algorithm: 0x%X\n", intermediateSignatureAlgorithm))
+	inputParams.WriteString(fmt.Sprintf("DC valid for: %v\n", dcValidFor))
+
+	inputParamsLog, err := os.Create(filepath.Join(testOutputsDir, "input-params.txt"))
+	if err != nil {
+		return err
+	}
+	defer inputParamsLog.Close()
+
+	_, err = inputParamsLog.WriteString(inputParams.String())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *testCaseDC) run(client endpoint, server endpoint) error {
```

Comment:
Do we expect the body of this to change across test cases? Nothing in here *seems* specific to the DC test case. If not, maybe we factor it out?

Reply:
Agree that nothing is DC-specific, and the body probably won't change, but I think it might be best to wait until we have a few more complicated test cases before refactoring this.